sdk/core/core-auth/src/tokenCredential.ts

```diff
@@ -49,11 +49,12 @@ export interface AccessToken {
  * @param credential The assumed TokenCredential to be tested.
  */
 export function isTokenCredential(credential: any): credential is TokenCredential {
-  // Check for an object with a 'getToken' function but without
+  // Check for an object with a 'getToken' function and possibly with
   // a 'signRequest' function.  We do this check to make sure that
   // a ServiceClientCredentials implementor (like TokenClientCredentials
-  // in ms-rest-nodeauth) doesn't get mistaken for a TokenCredential.
+  // in ms-rest-nodeauth) doesn't get mistaken for a TokenCredential if
+  // it doesn't actually implement TokenCredential also.
   return credential
     && typeof credential.getToken === "function"
-    && credential.signRequest === undefined;
+    && (credential.signRequest === undefined || credential.getToken.length > 0);
```

Comment:
What is the scenario where `credential.signRequest === undefined` will be true but `credential.getToken.length > 0` will be `false`?

Reply:
The logic here is "if `credential` has a `signRequest` method, `credential.getToken` must accept at least 1 parameter".  The point is to only check the number of parameters to `getToken` if we think the credential may also be a `ServiceClientCredential/TokenClientCredential`.  Brian had raised a concern about _only_ using the length check in the [previous PR](https://github.com/Azure/azure-sdk-for-js/pull/4335#discussion_r304174737).

Reply:
For the second question, `credential.signRequest === undefined` being `true` and `credential.getToken.length > 0` being `false` would happen when someone tries to pass a non-compatible `ms-rest-nodeauth` credential to `isTokenCredential` (one from a release before the changes I'm making in the other PR).

Reply:
My apologies, I didnt read https://github.com/Azure/azure-sdk-for-js/pull/4367#issuecomment-513902599 before commenting above :(